include/aspect/particle/property/cpo.h

```diff
@@ -0,0 +1,361 @@
+
+
+          /**
+           * Stores information in variables into the data array. An extention of load_particle_data which includes the grain
+           * derivatvies. Intended mostly for use in this plugin.
+           */
+          void
+          store_particle_data_extended(unsigned int cpo_data_position,
+                                       const ArrayView<double> &data,
+                                       std::vector<unsigned int> &deformation_type,
+                                       std::vector<double> &volume_fraction_mineral,
+                                       std::vector<std::vector<double>> &volume_fractions_mineral,
+                                       std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral,
+                                       std::vector<std::vector<double> > &volume_fractions_mineral_derivatives,
+                                       std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral_derivatives) const;
+
+          /**
+           * Updates the volume fractions and rotation matrices with a Forward Euler scheme.
+           */
+          double
+          advect_forward_euler(std::vector<double> &volume_fractions,
+                               std::vector<Tensor<2,3> > &a_cosine_matrices,
+                               const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                               const double strain_rate_second_invariant,
+                               const double dt) const;
+          /**
+           * Updates the volume fractions and rotation matrices with a Backward Euler scheme.
+           */
+          double
+          advect_backward_euler(std::vector<double> &volume_fractions,
+                                std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+          /**
+           * Updates the volume fractions and rotation matrices with a Crank-Nicolson scheme.
+           */
+          double
+          advect_Crank_Nicolson(std::vector<double> &volume_fractions,
+                                std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                std::vector<double> &previous_volume_fraction_derivatives,
+                                std::vector<Tensor<2,3> > &previous_a_cosine_matrices_derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+
+          /**
+           * A interface function which delegates the computation the derivatives
```

Comment:
You explain *how* this function does something, but it is more important to explain *what* this function does. Which derivatives? The fact that it calls other functions does not really matter.

```suggestion
           * An interface function, which computes and returns the derivatives of ...
```